pymc3/distributions/bound.py

```diff
@@ -0,0 +1,232 @@
+        Lower bound of the distribution, set to -inf to disable.
+    upper : float (optional)
+        Upper bound of the distribibution, set to inf to disable.
+    tranform : 'infer' or object
+        If 'infer', infers the right transform to apply from the supplied bounds.
+        If transform object, has to supply .forward() and .backward() methods.
+        See pymc3.distributions.transforms for more information.
+    """
+
+    def __init__(self, distribution, lower, upper,
+                 transform='infer', *args, **kwargs):
+        dtype = kwargs.get('dtype', theano.config.floatX)
+
+        if lower is not None:
+            lower = tt.as_tensor_variable(lower).astype(dtype)
+        if upper is not None:
+            upper = tt.as_tensor_variable(upper).astype(dtype)
+
+        if transform == 'infer':
+            if lower is None and upper is None:
+                transform = None
+                default = None
+            elif lower is not None and upper is not None:
+                transform = transforms.interval(lower, upper)
+                default = 0.5 * (lower + upper)
+            elif upper is not None:
+                transform = transforms.upperbound(upper)
+                default = upper - 1
+            else:
+                transform = transforms.lowerbound(lower)
+                default = lower + 1
+        else:
+            default = None
+
+        super(_ContinuousBounded, self).__init__(
+            distribution=distribution, lower=lower, upper=upper,
+            transform=transform, default=default, *args, **kwargs)
+
+
+class Bound(object):
+    R"""
+    Create a new upper, lower or upper+lower bounded distribution.
+
+    The resulting distribution is not normalized anymore. This
+    is usually fine if the bounds are constants. If you need
+    truncated distributions, use `Bound` in combination with
+    a `pm.Potential` with the cumulative probability function.
+
+    The bounds are inclusive for discrete distributions.
+
+    Parameters
+    ----------
+    distribution : pymc3 distribution
+        Distribution to be transformed into a bounded distribution.
+    lower : float or array like, optional
+        Lower bound of the distribution.
+    upper : float or array like, optional
+        Upper bound of the distribution.
+
+    Example
+    -------
+    # Bounded distribution can be defined before the model context
+    PositiveNormal = pm.Bound(pm.Normal, lower=0.0)
+    with pm.Model():
+        par1 = PositiveNormal('par1', mu=0.0, sd=1.0, testval=1.0)
+        # or within the model context
+        NegativeNormal = pm.Bound(pm.Normal, upper=0.0)
+        par2 = NegativeNormal('par2', mu=0.0, sd=1.0, testval=1.0)
+
+        # or you can define it implicitly within the model context
+        par3 = pm.Bound(pm.Normal, lower=-1.0, upper=1.0)(
+                'par3', mu=0.0, sd=1.0, testval=1.0)
+    """
+
+    def __init__(self, distribution, lower=None, upper=None):
```

Comment:
I'd rather not make this any more complicated than it already is.
Sadly, we don't distinguish properly between random variables and distributions, but use the constructor of our distribution objects to build random variables (observed and otherwise). This leads to the somewhat strange behaviour
```python
issubclass(pm.Normal, pm.Distribution)  # -> True
isinstance(pm.Normal('a'), pm.Distribution) # -> False
isinstance(pm.Normal('a'), pm.Normal) # -> False
```
As we really don't want to second to be true, I think the first actually *shouldn't* be true in the first place. We shouldn't think of our distribution classes as distributions, but as factory functions for random variables. The distributions are `pm.Normal.dist`.
So I kind of don't see the point of adding another non-standard thing on top of the metaclasses and overriden `__new__` that's already in there, to make a subclass check true, that shouldn't be true in the first place. Writing it isn't the problem, it just makes the code ever harder to read.

We do have the following now, which I think is the important part:
```python
BoundNormal = pm.Bound(pm.Normal)
isinstance(BoundNormal.dist(), pm.Continuous) # -> True
```

The original design of `pm.Bound` is I guess a bit unfortunate, we wouldn't have that problem in the first place if it took a distribution as parameter and returned a random variable:
```python
dist = pm.Normal.dist(mu=5, sd=3)
pm.Bound("a", dist, lower=3)
```